File: exercism/book-store/book_store.zig
const std = @import("std");
fn calculate(counts: []const u32) u32 {
    var ret: f32 = 0.0;
    var used: u32 = 0;
    const base_price = 800.0;
    const n = counts.len;
    for (0..n) |i| {
        const m = n - i;
        const x = counts[i] - used;
        used = counts[i];
        const price = base_price * @as(f32, @floatFromInt(m)) * @as(f32, @floatFromInt(x));
        const discount: f32 = switch (m) {
            2 => 0.95,
            3 => 0.90,
            4 => 0.80,
            5 => 0.75,
            else => 1.00,
        };
        ret += price * discount;
    }
    return @as(u32, @intFromFloat(ret));
}

pub fn total(basket: []const u32) u32 {
    var counts = std.mem.zeroes([5]u32);
    for (basket) |book| {
        counts[book - 1] += 1;
    }
    std.mem.sort(u32, &counts, {}, comptime std.sort.asc(u32));
    const delta = @min(counts[0], counts[2] - counts[1]);
    counts[0] -= delta;
    counts[1] += delta;
    return calculate(&counts);
}

File: solutions/go/book-store/3/book_store.go
package bookstore

import (
	"fmt"
	"sort"
    "math"
)

func calculate(counts []int) int {
	const BASE_PRICE = 800.0
	//fmt.Printf("solving with counts = %v\n", counts)
	n := 0
	ret := float32(0.0)
	fmt.Printf("counts %v\n", counts)
	for i, x := range counts {
		m := len(counts) - i
		x -= n
		n += x
		fmt.Printf("there are %d books in each %d kinds\n", x, m)
		price := BASE_PRICE * float32(x) * float32(m)
		switch m {
		case 2:
			price *= 0.95
		case 3:
			price *= 0.9
		case 4:
			price *= 0.8
		case 5:
			price *= 0.75
		default:
		}
		ret += price
		fmt.Printf("price %f\n", ret)
	}
	return int(ret)
}
func min(x, y int) int {
  return int(math.Min(float64(x), float64(y)))
}
func Cost(books []int) int {
	counts := make([]int, 5)
	for _, book := range books {
		counts[book-1]++
	}
	sort.Ints(counts)
	delta := min(counts[0], counts[2]-counts[1])
	counts[0] -= delta
	counts[1] += delta
	return calculate(counts)
}